### PR TITLE
[WIP] A way to submit tests that should fail.

### PR DESCRIPTION
**Pull request by cgcgbcbc**
see #673

### CODE REVIEW

**Comment by jfmengels** [comment]
There needs to be a test that makes the tests fail when a test passes.

**Comment by jamestalmage** [comment]
Agreed, needs quite a few more tests.

- [x] failing sync test fails - result is passing
- [x] failing test.cb test fails - result is passing
- [ ] failing test with `t.throws(nonThrowingPromise)` - result is passing (test async assertions)
- [ ] failing test returns a rejected promise - result is passing
- [ ] failing sync test passes - result is a failure with helpful error message
- [ ] failing test.cb test passes - result is a failure
- [ ] failing test with `t.throws(throws)` - result is failure (test async assertions)
- [ ] failing test returns a resolved promise - result is failure

**Comment by cgcgbcbc** [comment]
> failing test with t.throws(throws) - result is failure (test async assertions)

Shouldn't this be passing?

**Comment by cgcgbcbc** [comment]
Just for tracking

- [ ] failing sync test fails - result is passing
- [ ] failing test.cb test fails - result is passing
- [ ] failing test with t.throws(nonThrowingPromise) - result is passing (test async assertions)
- [ ] failing test returns a rejected promise - result is passing
- [ ] failing sync test passes - result is a failure with helpful error message
- [ ] failing test.cb test passes - result is a failure
- [ ] failing test with t.throws(throws) - result is failure (test async assertions)
- [ ] failing test returns a resolved promise - result is failure

**Comment by jamestalmage** [comment]
> Just for tracking

Any reason you haven't tackled the last two?

```js
// this should pass
test.failing(t => {
  t.throws(Promise.resolve('foo'));
});
// this should fail
test.failing(t => {
  t.notThrows(Promise.resolve('foo'));
});

**Comment by cgcgbcbc** [comment]
> Any reason you haven't tackled the last two?

Cause they failed...

**Comment by jamestalmage** [comment]
This should fix your failing test:

```diff
diff --git a/test/test.js b/test/test.js
index d2f3f1b..9fd23a9 100644
--- a/test/test.js
+++ b/test/test.js
@@ -676,10 +676,10 @@ test('failing test with t.throws(nonThrowingPromise) is passing', function (t) {
 
 test('failing test with t.throws(throws) is failure', function (t) {
 	ava.failing(function (a) {
-		a.throws(Promise.resolve('foo'));
+		a.notThrows(Promise.resolve('foo'));
 	}).run().then(function (result) {
 		t.is(result.passed, false);
-		t.is(result.message, failingTestHint);
+		t.is(result.reason.message, failingTestHint);
 		t.end();
 	});
 });
```

**Comment by cgcgbcbc** [comment]
😅Aha, not the first time making such embarrassing mistake

**Comment by jamestalmage** [comment]
Awesome!

LGTM!

**Comment by jamestalmage** [comment]
@cgcgbcbc - Can you add documentation for this to the Readme?

**Comment by cgcgbcbc** [comment]
@jamestalmage , do you think it's necessary to check that the failing modifier can only be used with test and throws error when used with hooks?

**Comment by jamestalmage** [comment]
> do you think it's necessary to check that the failing modifier can only be used with test and throws error when used with hooks?

No, it's conceivable you could write an `afterEach` test that failed.

```js
test.failing.afterEach('always leaves X in pristine state', t => {
  t.true(isPristine(t.context.state));
});
```

That's enough of a corner case that it should probably be flagged by the linter though.

Can you open an issue in `eslint-plugin-ava` about that?

**Comment by cgcgbcbc** [comment]
> Can you open an issue in eslint-plugin-ava about that?

Do you mean the no-unknown-modifiers rule?

**Comment by jamestalmage** [comment]
I just did it: https://github.com/sindresorhus/eslint-plugin-ava/issues/108

File: readme.md [hunk 1]
```diff
@@ -415,6 +415,16 @@ You can use the `.todo` modifier when you're planning to write a test. Like skip
 test.todo('will think about writing this later');
 ```
 
+### Failing tests
+
+You can use the `.failing` modifier for submitting tests that are intended to be failures without killing CI process. And if the failing test does pass, it will fail, and it will remind you remove the failing modifier.
```
Comment (jamestalmage) [inline readme.md]: I would like to expand on this a bit: 

> You can use the `.failing` modifier to document issues with your code that need to be fixed. Failing tests are run just like normal ones, but they are expected to fail, and will not break your build when they do. If a test marked as `failing` actually passes, it will be reported as an error and fail the build with a helpful message instructing you to remove the `.failing` modifier.

> This allows you to merge `.failing` tests before a fix is implemented without breaking CI. This is a great way to recognize good bug report PR's with a commit credit, even if the reporter is unable to actually fix the problem.

File: readme.md [hunk 2]
```diff
@@ -415,6 +415,16 @@ You can use the `.todo` modifier when you're planning to write a test. Like skip
...
+
+```js
+test.failing('this test prove something wrong', t => {
```
Comment (jamestalmage) [inline readme.md]: People should link to the relevant issue. We should model that behavior in the docs.

Also, a bit of language/style cleanup.

```js
// See: github.com/user/repo/issues/1234
test.failing('demonstrate some bug', t => {
  t.fail(); // test will count as passed
});
```

**Comment by cgcgbcbc** [comment]
@jamestalmage Thank you for your advice! I've fixed the documents.

**Comment by jamestalmage** [comment]
This looks great to me!

Just need a second reviewer to merge.

// @sindresorhus @novemberborn @vdemedes @sotojuan

**Comment by novemberborn** [comment]
Looks good @cgcgbcbc!

I have some nits but will follow up with a PR rather than putting you through more back-and-forth 😄

**Comment by sotojuan** [comment]
Thanks @cgcgbcbc!
